week-3/mandatory/2-api/server.js

```diff
@@ -0,0 +1,307 @@
+const express = require("express");
+const app = express();
+const db = require("./db");
+app.use(express.json());
+
+// ---- Welcome Page-----
+
+app.get("/", (req, res) => {
+  res.send(`<h1>Wellcome to CYF Ecommerce API<h1>
+  <h3>You can query ...</h3>
+  <h4>all customers:  '/customers' <br>
+  with id:  '/customers/id_no' <br>
+  with string search in names:  '/customers/by_name/str' <h4> 
+  <h4>You can also query suppliers '/suppliers' or products '/products'</h4> `);
+});
+
+// ------ Get All Customers-----------
+
+app.get("/customers", function (req, res) {
+  db.query(`SELECT * FROM customers`)
+    .then((result) => {
+      if (result) res.json(result.rows);
+    })
+    .catch((err) => res.status(500).json({ error: err }));
+});
+
+// --------- Get All Suppliers -----------
+
+app.get("/suppliers", function (req, res) {
+  db.query(`SELECT * FROM suppliers`, (error, result) => {
+    if (!error) res.json(result.rows);
+  });
+});
+
+// -----------Get Customers by ID-------------
+
+app.get("/customers/:id", function (req, res) {
+  let custId = Number(req.params.id);
+  db.query("SELECT * FROM customers WHERE id = $1", [custId], function (
+    error,
+    result
+  ) {
+    if (!error) res.json(result.rows[0]);
+  });
+});
+
+// ------------Filter Customers by Name search--------
+
+app.get("/customers/by_name/:name", (req, res) => {
+  let custName = req.params.name.toLowerCase();
+  db.query(
+    "SELECT * FROM customers WHERE LOWER(name) LIKE '%'||$1||'%'",
+    [custName],
+    (err, result) => {
+      if (err == undefined) {
+        res.status(200).json({ customers: result.rows });
+      } else {
+        console.log(err);
+        res.status(400).json(err);
+      }
+    }
+  );
+});
+
+//  Get Products & filtered Products on the same path---
+
+// ---- Solution 1------------
+
+app.get("/products", function (req, res) {
+  const name = req.query.name;
+
+  let allProducts = `SELECT products.product_name, pa.unit_price price, sup.supplier_name
+  FROM products JOIN product_availability pa ON (products.id=pa.prod_id)
+  JOIN suppliers sup ON (pa.supp_id=sup.id) `;
+
+  let filteredProducts =
+    allProducts + " WHERE LOWER(products.product_name) LIKE '%'||$1||'%'";
+
+  !name
+    ? db.query(allProducts, [], (error, result) => {
+        if (!error) res.json(result.rows);
+      })
```

Comment:
Naming is a bit weird, you can call it allProductQuery instead of allProducts
calling db on conditions is not ideal, you can do your checks on JS directly 
Something like 

```suggestion
sqlArgs=[]
      if (name){
allProduct+=filterporuct;
sqlArgs=[name]
db.query(allproduct, sqlargs...
}
```